PostHog/Replay/PostHogReplayIntegration.swift

```diff
@@ -129,33 +134,133 @@
-            let style = RRStyle()
 
-            // no parent id means its the root
-            if parentId == nil, config.sessionReplayConfig.screenshotMode {
-                if let image = view.toImage() {
-                    wireframe.base64 = imageToBase64(image)
+            return wireframe
+        }
+
+        private func findMaskableWidgets(_ view: UIView, _ parent: UIView, _ maskableWidgets: inout [CGRect]) {
+            if let textView = view as? UITextView {
+                if isTextViewSensitive(textView) {
+                    maskableWidgets.append(view.toAbsoluteRect(parent))
+                    return
+                }
+            }
+
+            if let textField = view as? UITextField {
+                if isTextFieldSensitive(textField) {
+                    maskableWidgets.append(view.toAbsoluteRect(parent))
+                    return
+                }
+            }
+
+            if let image = view as? UIImageView {
+                if isImageViewSensitive(image) {
+                    maskableWidgets.append(view.toAbsoluteRect(parent))
+                    return
+                }
+            }
+
+            if view is UILabel {
+                if isTextInputSensitive(view) {
+                    maskableWidgets.append(view.toAbsoluteRect(parent))
+                    return
+                }
+            }
+
+            if swiftUIImageTypes.contains(where: { view.isKind(of: $0) }) {
+                if isAnyInputSensitive(view) {
+                    maskableWidgets.append(view.toAbsoluteRect(parent))
+                    return
+                }
+            }
+
+            if !view.subviews.isEmpty {
+                for child in view.subviews {
+                    if !child.isVisible() {
+                        continue
+                    }
+
+                    findMaskableWidgets(child, parent, &maskableWidgets)
                 }
-                wireframe.type = "screenshot"
-                return wireframe
             }
+        }
+
+        private func toScreenshotWireframe(_ view: UIView) -> RRWireframe? {
+            if !view.isVisible() {
+                return nil
+            }
+
+            var maskableWidgets: [CGRect] = []
+            findMaskableWidgets(view, view, &maskableWidgets)
+
+            let wireframe = createBasicWireframe(view)
+
+            if let image = view.toImage() {
+                let redactedImage = UIGraphicsImageRenderer(size: image.size, format: .init(for: .init(displayScale: 1))).image { context in
+                    context.cgContext.interpolationQuality = .none
+                    image.draw(at: .zero)
+
+                    for rect in maskableWidgets {
+                        let path = UIBezierPath(roundedRect: rect, cornerRadius: 10)
+                        UIColor.black.setFill()
+                        path.fill()
+                    }
+                }
+                wireframe.base64 = imageToBase64(redactedImage)
```

Comment:
possible optimization for the future, this can be done async but requires changing a few things and adjusting timestamps, if performance is an issue, this is a place to look at.

Reply:
I've started doing this optimization but led me to do other things that also required to be in the main thread so the changes didn't make much effect, most likely we'll need to pay the memory footprint price instead of processing the images right away, so we can batch the process in background threads but pay the price of holding a copy of the images in memory which can lead to OOM quite quickly as well.